File: Crc32.java
// Crc32.java
import java.io.*;
import java.nio.file.*;
import java.util.zip.CRC32;
import java.util.zip.CRC32C;
import java.util.*;

public class Crc32 {
    public static void main(String[] args) throws Exception {
        boolean useCrc32c = false;
        String check = null;
        boolean dec = false, bin = false;
        List<String> inputs = new ArrayList<>();

        for (int i=0; i<args.length; i++) {
            if (args[i].equals("--crc32c")) {
                useCrc32c = true;
            } else if (args[i].equals("--check") && i+1 < args.length) {
                check = args[++i];
            } else if (args[i].equals("--dec")) {
                dec = true;
            } else if (args[i].equals("--bin")) {
                bin = true;
            } else {
                inputs.add(args[i]);
            }
        }

        String fmt = "hex";
        if (dec) fmt = "dec";
        else if (bin) fmt = "bin";

        // Проверка stdin
        if (inputs.isEmpty() && System.in.available() > 0) {
            ByteArrayOutputStream baos = new ByteArrayOutputStream();
            byte[] buf = new byte[8192];
            int n;
            while ((n = System.in.read(buf)) > 0) {
                baos.write(buf, 0, n);
            }
            byte[] data = baos.toByteArray();
            long hash = computeCrc32(data, useCrc32c);
            System.out.println(formatHash(hash, fmt));
            return;
        }

        if (inputs.isEmpty()) {
            System.err.println("Не указаны данные.");
            System.exit(1);
        }

        for (String item : inputs) {
            Path path = Paths.get(item);
            if (Files.exists(path) && !Files.isDirectory(path)) {
                byte[] data = Files.readAllBytes(path);
                long hash = computeCrc32(data, useCrc32c);
                String output = item + ": " + formatHash(hash, fmt);
                if (check != null) {
                    long expected = Long.parseLong(check.startsWith("0x") ? check.substring(2) : check, 16);
                    output += (expected == hash) ? " (✅ OK)" : " (❌ FAIL)";
                }
                System.out.println(output);
            } else {
                byte[] data = item.getBytes("UTF-8");
                long hash = computeCrc32(data, useCrc32c);
                String output = formatHash(hash, fmt);
                if (check != null) {
                    long expected = Long.parseLong(check.startsWith("0x") ? check.substring(2) : check, 16);
                    output += (expected == hash) ? " (✅ OK)" : " (❌ FAIL)";
                }
                System.out.println(output);
            }
        }
    }

    private static long computeCrc32(byte[] data, boolean useCrc32c) {
        if (useCrc32c) {
            CRC32C crc = new CRC32C();
            crc.update(data);
            return crc.getValue();
        } else {
            CRC32 crc = new CRC32();
            crc.update(data);
            return crc.getValue();
        }
    }

    private static String formatHash(long hash, String fmt) {
        if (fmt.equals("hex")) {
            return String.format("0x%08X", hash);
        } else if (fmt.equals("dec")) {
            return String.valueOf(hash);
        } else if (fmt.equals("bin")) {
            return String.format("%32s", Long.toBinaryString(hash)).replace(' ', '0');
        }
        return String.format("0x%08X", hash);
    }
}

File: crc32.cs
// crc32.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.IO.Hashing;
using System.Text;

class Crc32
{
    static void Main(string[] args)
    {
        bool useCrc32c = false;
        string check = null;
        bool dec = false, bin = false;
        var inputs = new List<string>();

        for (int i=0; i<args.Length; i++)
        {
            switch (args[i])
            {
                case "--crc32c": useCrc32c = true; break;
                case "--check": if (i+1 < args.Length) check = args[++i]; break;
                case "--dec": dec = true; break;
                case "--bin": bin = true; break;
                default: inputs.Add(args[i]); break;
            }
        }

        string fmt = "hex";
        if (dec) fmt = "dec";
        else if (bin) fmt = "bin";

        // Проверка stdin
        if (inputs.Count == 0 && Console.IsInputRedirected)
        {
            using (var ms = new MemoryStream())
            {
                Console.OpenStandardInput().CopyTo(ms);
                byte[] data = ms.ToArray();
                uint hash = ComputeCrc32(data, useCrc32c);
                Console.WriteLine(FormatHash(hash, fmt));
                return;
            }
        }

        if (inputs.Count == 0)
        {
            Console.Error.WriteLine("Не указаны данные.");
            Environment.Exit(1);
        }

        foreach (var item in inputs)
        {
            if (File.Exists(item))
            {
                byte[] data = File.ReadAllBytes(item);
                uint hash = ComputeCrc32(data, useCrc32c);
                string output = $"{item}: {FormatHash(hash, fmt)}";
                if (check != null)
                {
                    uint expected = Convert.ToUInt32(check.StartsWith("0x") ? check.Substring(2) : check, 16);
                    output += (expected == hash) ? " (✅ OK)" : " (❌ FAIL)";
                }
                Console.WriteLine(output);
            }
            else
            {
                byte[] data = Encoding.UTF8.GetBytes(item);
                uint hash = ComputeCrc32(data, useCrc32c);
                string output = FormatHash(hash, fmt);
                if (check != null)
                {
                    uint expected = Convert.ToUInt32(check.StartsWith("0x") ? check.Substring(2) : check, 16);
                    output += (expected == hash) ? " (✅ OK)" : " (❌ FAIL)";
                }
                Console.WriteLine(output);
            }
        }
    }

    static uint ComputeCrc32(byte[] data, bool useCrc32c)
    {
        if (useCrc32c)
        {
            return Crc32C.HashToUInt32(data);
        }
        else
        {
            return System.IO.Hashing.Crc32.HashToUInt32(data);
        }
    }

    static string FormatHash(uint hash, string fmt)
    {
        if (fmt == "hex") return $"0x{hash:X8}";
        if (fmt == "dec") return hash.ToString();
        if (fmt == "bin") return Convert.ToString(hash, 2).PadLeft(32, '0');
        return $"0x{hash:X8}";
    }
}
